blog: docker build alternatives (#976)
tilt-dev/tilt.build: blog/_posts/2021-11-04-docker-build-alternatives.md

---
slug: "docker-build-alternatives"
date: 2021-11-03
author: nick
layout: blog

> title: "How to Cope When Docker Builds are Slowing You Down"

Reviewer: I feel like "Docker builds slowing you down" restricts the problem space unnecessarily.
Instead, I'd suggest something like "When Docker Builds are not right for you" (or something better)

Author: Ya, i was trying to do a hat-tip to Jason's tweet below. But I think it needs a catchier title. let me bat some ideas around

Jason Hall recently tweeted about this, then wrote [a great blog
post](https://github.com/ImJasonH/ImJasonH/tree/main/articles/moving-and-building-images)
expanding on it.

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">You don&#39;t need `docker build` to build a container image, and you don&#39;t need to `docker pull / tag / push` to move one around. In a lot of cases it&#39;s actively slowing you down, or making you less secure.</p>&mdash; Jason is no longer spooky (@ImJasonH) <a href="https://twitter.com/ImJasonH/status/1446624521507819521?ref_src=twsrc%5Etfw">October 8, 2021</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>

Teams using Tilt often ask us: if my Docker build is getting slow, how can Tilt help me make it faster?

Fortunately, Tilt lets you swap out and experiment with different image builders
in dev. And you can package up those experiments as
[extensions](https://github.com/tilt-dev/tilt-extensions). The Tilt community
frequently adds new types of image builders.

In this post, I want to brag about three very different types of image builders that exist today:

- `kubectl_build`,

- `ko`, and

> - `buildpack`

Author: ```suggestion
- `pack`
```

  '.', 
  build_args={'flask_env': 'development'},
  live_update=[
    sync('.', '/app'),
  ])
```

The main difference is that you will now see a buildkit server in your cluster:

```shell
$ kubectl get deployment
NAME             READY   UP-TO-DATE   AVAILABLE   AGE
buildkit         1/1     1            1           70m
example-python   1/1     1            1           7s
```

This can be nice if you have a powerful cluster and a 13" laptop!

Read the full docs on the `kubectl_build` extension [here](https://github.com/tilt-dev/tilt-extensions/tree/master/kubectl_build).

## Ko

Maybe you're not happy with your Dockerfile.

The Dockerfile view of the world is: to build an image, run some commands in a container,
then save the results of those commands.

The Ko view is: an image is a tarball. If you have a Mac, why do you need a
Linux VM to run a table if you want a tarball? You already have good local tools
for making tarballs.

[Ko](https://github.com/google/ko#install) is an image builder targeted at Go
apps. Go apps are particularly ill-suited for containers. The Go toolchain is
well-optimized to run locally and builds a single binary. Ko builds your Go app

> locally, then packs it into an image.

Author: done

locally, then packs it into an image. No build containers or VMs needed!

A while back, I wrote a small extension that lets you use `ko` in Tilt. Here's what it looks like:

```
load('ext://ko', 'ko_build')

ko_build(
  'example-go-image',
  './',
  deps=['.']
)
```

The upside of `ko` is that it's super simple to get started.

The downside is that if you want any features that can't be expressed in a
static Go binary, you're going to have a difficult time. Not impossible, but
difficult. This includes: static files on-disk, dynamic C libraries, or
hot-reload.

Read the full docs on the `ko` extension
[here](https://github.com/tilt-dev/tilt-extensions/tree/master/ko).

## Pack

[`pack`](https://buildpacks.io/docs/tools/pack/) is also addressing people who
are not happy maintaining Dockerfiles.

But where Ko's philosophy states that you don't need a container to build an

> image, Buildpack takes the viewpoint that you need MORE containers.

Reviewer 2: I would say the philosophy is that you shouldn't have to care how an image is built, like if a tool is using more or fewer containers to make that happen. Give `pack` your source code, it'll give you an image. What happens in between are details that you as an app dev don't need to manage or care about.

Author: haha, yes, this was a bit tongue-and-cheek, but you're right that i should fold in the actual philosophy and not the joke one

Wait, wait, don't run away. Let me explain.

With `pack`, there's a stronger separation between the container you use
to build an image and the container you end up running. The container images you use
to build stuff are called buildpacks.

Separating these out has some big advantages!

- You can stack buildpacks to mix in the features you want (like hot-reload tools).

- The buildpack keeps out lots of junk you don't want in the final image (like compilers).

The `pack` extension was one of the first Tilt extensions! [Gareth
Rushgrove](https://github.com/garethr) added it when we were still experimenting
with the extension system.

Since then, we worked closely with [Daniel

> Makusa](https://github.com/dmikusa-pivotal) of the buildpacks team to make it

Reviewer 2: ```suggestion
Mikusa](https://github.com/dmikusa-pivotal) of the buildpacks team to make it
```

Author: oops, that's embarrassing, thank you

even better.

Here's what it looks like in a Tiltfile:

```python
pack(
  'example-python-image', 
  '.',
  buildpacks=[
    'gcr.io/paketo-buildpacks/python',
    'gcr.io/paketo-buildpacks/environment-variables',
  ],
  env_vars=['BPE_FLASK_ENV=development'],
  live_update=[
    sync('.', '/workspace'),
  ])
```

In this example, we're using two buildpacks: a `python` buildpack that sets up a
Python image, and a `environment-variables` buildpack that sets
`FLASK_ENV=development` to get live-reloading (Flask is the python framework
that does the live-reload.)

Read the full docs on the `pack` extension
[here](https://github.com/tilt-dev/tilt-extensions/tree/master/pack).

### Fun aside: Future Buildpack Work

The layering approach to adding new functionality is getting a lot of excitement
right now from container nerds.

For security nerds, you can better track where source files come from and attach
SBOMs (which are way beyond the scope of this post, but are fun to say out
load.)

> One of the big reasons Tilt nerds are excited about buildpacks is that there's a

Reviewer: ```suggestion
One of the big reasons Tilters are excited about buildpacks is that there's a
```
I feel like you are using the term "nerds" a lot, you might want to switch it up once in a while 🤓
